src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Extension.cs

```diff
@@ -63,6 +64,309 @@ internal override string ExtensionName
             }
         }
 
+        /// <summary>
+        /// This name uses an IL-looking format to encode CLR-level information of an extension block (ie. arity, constraints, extended type).
+        /// It is meant be to hashed to produce the content-based name for the extension grouping type.
+        /// </summary>
+        internal string ComputeExtensionGroupingRawName()
+        {
+            Debug.Assert(this.IsExtension && this.IsDefinition);
+
+            var pooledBuilder = PooledStringBuilder.GetInstance();
+            var builder = pooledBuilder.Builder;
+            builder.Append("extension");
+
+            if (this.Arity > 0)
+            {
+                builder.Append("<");
+
+                foreach (var typeParameter in this.TypeParameters)
+                {
+                    if (typeParameter.Ordinal > 0)
+                    {
+                        builder.Append(", ");
+                    }
+
+                    appendTypeParameterDeclaration(typeParameter, builder);
+                }
+
+                builder.Append(">");
+            }
+
+            builder.Append("(");
+            if (this.ExtensionParameter is { } extensionParameter)
+            {
+                appendType(extensionParameter.Type, builder);
+            }
+
+            builder.Append(")");
+
+            return pooledBuilder.ToStringAndFree();
+
+            static void appendType(TypeSymbol type, StringBuilder builder)
+            {
+                if (type is NamedTypeSymbol namedType)
+                {
+                    appendNamedType(type, builder, namedType);
+                }
+                else if (type is TypeParameterSymbol typeParameter)
+                {
+                    appendTypeParameterReference(typeParameter, builder);
+                }
+                else if (type is ArrayTypeSymbol array)
+                {
+                    appendArrayType(array, builder);
+                }
+                else if (type is PointerTypeSymbol pointer)
+                {
+                    appendType(pointer.PointedAtType, builder);
+                    builder.Append('*');
+                }
+                else if (type is FunctionPointerTypeSymbol functionPointer)
+                {
+                    appendFunctionPointerType(functionPointer, builder);
+                }
+                else
+                {
+                    throw ExceptionUtilities.UnexpectedValue(type);
+                }
+            }
+
+            static void appendNamedType(TypeSymbol type, StringBuilder builder, NamedTypeSymbol namedType)
+            {
+                Debug.Assert(type.CustomModifierCount() == 0);
+
+                if (namedType.SpecialType == SpecialType.System_Void)
+                {
+                    builder.Append("void");
```

Comment:
>builder.Append("void"); [](http://example.com/codeflow?start=20&length=23)

It looks like other special types also have keyword-like names (see I.8.2.2 Built-in value and reference types). Consider using them as well, at least for those types that have a dedicated entry in "I I.23.1.16 Element types used in signatures" table.